### PR TITLE
Use ITextBufferCloneService if available in AsText

### PR DESCRIPTION
**Pull request by sharwell**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/604150

/cc @minestarks

### Customer scenario

A customer attempts to reformat an HTML document, and Visual Studio crashes. Pasting the following into an HTML document will reproduce the issue.

```html
<html lang="en" xmlns="http://www.w3.org/1999/xhtml">
<head>
    <meta charset="utf-8" />
    <title></title>
    <script>    
        function foo() {}
    </script>
</head>
<body>

</body>
</html>
```

### Bugs this fixes

https://devdiv.visualstudio.com/DevDiv/_workitems/edit/604150

### Workarounds, if any

None.

### Risk

Low. Of multiple possible places to fix the sequence of events leading to a product crash, this one provides the strongest assurances that 15.7 release will continue to behave like 15.6 w.r.t. text differencing.

### Performance impact

Aside from fixing a possible crash, performance is improved in the impacted scenarios by allowing the use of `ITextBufferCloneService` rather than falling back to the slower path.

### Is this a regression from a previous update?

Yes, introduced by #25558.

### Root cause analysis

Two changes in the causal PR resulted in a behavior change:

* The `AsText(ITextSnapshot)` extension method failed to use the `ITextBufferCloneService` assigned to the source text buffer for the snapshot.
* Roslyn only uses text buffers created by `ITextBufferFactoryService`. The TypeScript implementation, on the other hand, uses projection buffers created by `IProjectionBufferFactoryService`. The events notifying extension code that a buffer is created differs for these two services.

While the fallback behavior of Roslyn was intended to be slower but functionally equivalent, the text differencing algorithm differed between the two approaches - the one used when `ITextBufferCloneService` was not available threw an exception if the set of changes made to a buffer were not ordered.

### How was the bug found?

Dogfooding / reported by TypeScript team

### Test documentation updated?

N/A

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 1]
```diff
@@ -1096,6 +1100,7 @@ protected override void Dispose(bool finalize)
             if (!finalize)
             {
                 _textBufferFactoryService.TextBufferCreated -= AddTextBufferCloneServiceToBuffer;
+                _projectionBufferFactoryService.ProjectionBufferCreated -= AddTextBufferCloneServiceToBuffer;
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: so we do this for every buffer or only roslyn buffer?

**Comment by jinujoseph** [comment]
approved to merge via [link](https://github.com/dotnet/roslyn/pull/26339)

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs (hunk 1 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: ➡️ All of them

**Comment by jaredpar** [comment]
>             e.TextBuffer.Properties.AddProperty(typeof(ITextBufferCloneService), _textBufferCloneService);

Not for this change but in general we should really avoid the use of Type instances as key in the property bags. Invites collisions with other services.

---
Refers to: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:229 in ab34327. [](commit_id = ab34327e599587188d423573a007ce9b7ddfcc9d, deletion_comment = False)

**Pull request merged**